### ISSUE TITLE
More and proper tags for observation spans

### ISSUE THREAD

**Issue by cmergenthaler** [spring-projects/spring-kafka#2609]
**Expected Behavior**

The current tags for tracing spans are very limited and contain only a few relevant information. [Opentelemetry](https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/messaging/) defines all the recommended and required attributes a single message span should include. It would be nice to have them predefined by spring-kafka automatically instead of using the customizer explicitly. Even though customizing might work, we lack a few information on the customizer to follow otel standardization properly (used broker, consumergroup, partition for producer, etc.).

**Current Behavior**

The only tags currently included are spring related tags (e.g. `spring.kafka.template`) and do not follow any standard. There are other spring projects that already have proper otel tags defined (e.g. [spring-data-mongodb](https://github.com/spring-projects/spring-data-mongodb/blob/main/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/observability/MongoObservation.java)). 

I think it would be beneficial to have consistent standard support over different spring projects. 

**Context**

To workaround this one can implement their own KafkaTemplateObservationConvention and define the necessary tags:

```java
@Component
public class KafkaTemplateObservationConvention extends KafkaTemplateObservation.DefaultKafkaTemplateObservationConvention {

    @Override
    public KeyValues getLowCardinalityKeyValues(KafkaRecordSenderContext context) {
        context.setRemoteServiceName(context.getBroker()); //peer.service -> should be the broker which is not available here
        return KeyValues.of(KafkaTemplateObservation.TemplateLowCardinalityTags.BEAN_NAME.asString(),
                context.getBeanName())
                .and("messaging.system", "kafka")
                .and("messaging.destination.kind", "topic")
                //.and("messaging.kafka.partition", context.getCarrier().partition().toString()) //Partition is always null
                //.and("messaging.kafka.message.offset", context.getCarrier().offset()) //Offset is not exposed
                .and("messaging.destination.name", context.getDestination());
    }

    @Override
    public String getContextualName(KafkaRecordSenderContext context) {
        return context.getDestination() + " publish";
    }
}
```

As you can see some required/recommended information is missing on the customizer.

**Comment by cmergenthaler** [comment]
I'll create a PR for that soon :)

**Comment by artembilan** [comment]
>  `return context.getDestination() + " publish";`

I believe previously it was `send`...
That's why we have it like that everywhere.

**Comment by cmergenthaler** [comment]
No problem, I will just change that. 

Do you know if it is possible to get the name of the broker (+ port) for the consumed/produced message somehow?

**Comment by artembilan** [comment]
Well, no problem with `publish`!
We may accept that change as well.
Not sure if we should treat it as a breaking change since even that OpenTelemetry does not do that when they rename attributes or their recommended values.

Re. broker name. Quoting that OTEL docs:

> For Apache Kafka producers, [peer.service](https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/span-general/#general-remote-service-attributes) SHOULD be set to the name of the broker or service the message will be sent to. The service.name of a Consumer’s Resource SHOULD match the peer.service of the Producer, when the message is directly passed to another service.

It might be great if they give us some hint what they talk about.
There is no something like plain "broker" in Kafka: we always connect to some cluster.
See respective `bootstrap.servers` Kafka property to connect: https://kafka.apache.org/documentation/#producerconfigs_bootstrap.servers.

So, probably we just can take that property from config.
However according to their docs it is something arbitrary chosen by end-user: https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/span-general/#general-remote-service-attributes

**Comment by artembilan** [comment]
> Instrumentations SHOULD provide a way for users to configure this name

Right, so, that's probably what a `KafkaTemplateObservationConvention` is for.

**Comment by jonatan-ivanov** [comment]
I would like to call out something if I may:
Most of the OTel semantic conventions are not stable yet (see "Status: Experimental" in the linked doc above). This means that they can change in the future: keys can be renamed, like `messaging.message.abc` -> `messaging.message.xyz`.

I think adding extra data is a great idea but we should be comfortable enough with the names to keep them even if OTel changes the specifications.

**Comment by jonatan-ivanov** [comment]
Sounds good!
Also, the `ObservationConvention` concept can help a lot in the future if needed.
Let's say we start using the OTel messaging conventions, release it then the convention will change in a breaking way and later becomes stable. If there will be a high enough user need, a secondary `ObservationConvention` can be added and users can choose if they want to use the default one that follows the older conventions (not breaking them) or the secondary one that follows the newer ones (breaking them).

**Comment by garyrussell** [comment]
I suggested something similar here https://github.com/spring-projects/spring-kafka/pull/2616#discussion_r1137708490

Just provide OTEL convention implementations that the user can choose to use instead of the default ones.

**Comment by artembilan** [comment]
Sure! 

That's your call.
I'll be happy to review whatever you end up.
